### PR TITLE
[Build] Update Build/Depends systems from upstream

### LINKED ISSUE #707: Bump minimum supported MacOS version to 10.10
Currently we support MacOS (OS X) down to version [10.8](https://github.com/pivx-project/pivx/blob/master/depends/hosts/darwin.mk#L1), which has worked fine so far but has required some code work arounds.

However a planned update to the version of Qt we use for release builds (Qt 5.9.6), which only supports MacOS [10.10+](https://wiki.qt.io/Qt_5.9_Tools_and_Versions#osx_10.10_x86_64), means that users will need to update their OS version to continue using the wallet.

At the time of writing this, all Mac computers that are capable of running 10.8 are also capable of running 10.10, and updates are free of charge. Because of this, I don't see any reason to hold back the Qt update or to continue supporting MacOS versions prior to 10.10.

For Reference:
- MacOS (OS X) 10.10 was released in 2014, with official support ending in 2017 (Security updates **MAY** still be released, but unlikely).
- Qt 5.9 update resolves a long running issue of needing to compile separate binaries for MacOS High Sierra, which will result in once again having only a single download file that covers the entire range of supported MacOS versions.

### PR DESCRIPTION
**Pull request by Fuzzbawls**
This updates a number of underlaying/supporting systems from upstream. Of particular notability is the update to Qt5.9 and the move to using Ubuntu Bionic for TravisCI and Gitian builds.

- Updates Depends dependencies to recent versions, including Qt5.9 (closes #707 and #364).
- Updates the `configure` script and dependent makefiles with upstream declarations.
- Resolves minor conflicts with our GMP/OpenSSL detections.
- Lays down new basic (but as yet unimplemented) groundwork for lcov coverage reporting.
- Gitian descriptor updates to build on Ubuntu Bionic instead of the dated Ubuntu Trusty.
- Re-unify the linux and aarch64 gitian descriptors now that Ubuntu Bionic fixes cross compilation for all supported linux arches.
- Default Symbol/Security checking is now performed for gitian builds (Linux ABI compatibility).
- TravisCI updates to provide cleaner log output, and also utilize Ubuntu Bionic via Docker.

This PR depends on #752, and as such will fail the TravisCI test for the MacOS target. I will rebase once that PR is merged.

### CODE REVIEW

**Comment by bumbacoin** [comment]
in the interest of checking for macOSX build compatibility
i pulled the code here as is,

succesfully built using ubuntu 16 and 10.11sdk  on ubuntu16

with the wallet successfully opening on High Sierra running VM in virtualbox,
and Yosemite running natively on a macbook

there were some strange graphic artifacts in High Sierra, possibly from the VM.

**Comment by flyinghuman** [comment]
@bumbacoin have your wallet a backgound image? i get the same artifacts inside VM on Sierra and Mojave but only when the window has a background-image - not tested on real mac so far.

**Comment by Fuzzbawls** [comment]
Thus far, testing on a native High Sierra capable mac have shown no visual discrepancies.

**Comment by bumbacoin** [comment]
here's a copy of image showing glitches in VM

none are visible in send receive and privacy tabs

the pink lines appear minimally in the overview and masternode tabs
pivx-graphicalglitch
and maximally in the receive tab
pivx-graphicalglitch2

**Comment by MitchellCash** [comment]
I also see these same pink artefacts on MacBook Pro High Sierra after a Gitian build. So no VM.

They seem intermittent, sometimes it loads okay. But there is definitely an issue somewhere.

**Comment by bumbacoin** [comment]
i tested last official high sierra build on the same vm and there are no graphical glitches.

**Comment by Fuzzbawls** [comment]
Give the .dmg file from https://www.fuzzbawls.pw/pivx/testnet a test to see if the graphical glitches are present

**Comment by bumbacoin** [comment]
> Give the .dmg file from https://www.fuzzbawls.pw/pivx/testnet a test to see if the graphical glitches are present

this shows the glitches

**Comment by flyinghuman** [comment]
"Problem" is the
`
border-image: url(':/images/walletFrame_bg') 0 0 0 0 stretch stretch;
`
inside default.css; seems that this is rendered different in QT5.9 than in previous versions.

**Comment by flyinghuman** [comment]
i changed it to:
```
background-image: url(':/images/walletFrame_bg');
background-size: 100% 100%;
background-position: bottom right;
background-repeat: no-repeat;
background-attachment: fixed;
```
but QT seems to not understand background-size. For me its okay ATM. The Glitches are gone.

**Comment by flyinghuman** [comment]
using bionic leads to newer libstdc++6 library - i think this is the reason for this error here:
```
/usr/include/c++/7/debug/safe_iterator.h:270:
Error: attempt to dereference a past-the-end iterator.

Objects involved in the operation:
    iterator "this" @ 0x0x7ffd73f035b0 {
      type = gnu_debug::_Safe_iterator<std::_Rb_tree_iterator<std::pair<uint256 const, CMasternodeBroadcast> >, std::debug::map<uint256, CMasternodeBroadcast, std::less<uint256>, std::allocator<std::pair<uint256 const, CMasternodeBroadcast> > > > (mutable iterator);
      state = past-the-end;
      references sequence with type 'std::__debug::map<uint256, CMasternodeBroadcast, std::less<uint256>, std::allocator<std::pair<uint256 const, CMasternodeBroadcast> > >' @ 0x0x55a98bf6ae20
    }
```
Before changing dependencies this error did not happen to me. Can anyone give me a hint to resolve this issue? It happend running on Ubuntu 16.04 with testing libstdc++ libs from apt-repository ppa:ubuntu-toolchain-r/test to get on the actual version for 16.04:
```
add-apt-repository ppa:ubuntu-toolchain-r/test
apt-get update && apt-get install libstdc++6
```
Thanks!

File: depends/README.md [hunk 1]
```diff
@@ -22,34 +22,51 @@ Common `host-platform-triplets` for cross compilation are:
...
+
+For linux AARCH64 cross compilation:
+
+    sudo apt-get install g++-aarch64-linux-gnu binutils-aarch64-linux-gnu
+
+For linux RISC-V 64-bit cross compilation (there are no packages for 32-bit):
+
+    sudo apt-get install g++-riscv64-linux-gnu binutils-riscv64-linux-gnu
+
+RISC-V known issue: gcc-7.3.0 and gcc-7.3.1 result in a broken `test_bitcoin` executable (see https://github.com/bitcoin/bitcoin/pull/13543),
```
[the remark this answers is not in the document]
Comment (Mrs-X) [reply depends/README.md]: @Fuzzbawls I think it should indeed be `test_pivx`,  the configure/makefiles all refer to this.

[the remark this answers is not in the document]
Comment (Fuzzbawls) [reply depends/README.md]: @Warrows @Mrs-X fixed the typo.

**Comment by Fuzzbawls** [comment]
> using bionic leads to newer libstdc++6 library - i think this is the reason for this error here:
> 
> ```
> /usr/include/c++/7/debug/safe_iterator.h:270:
> Error: attempt to dereference a past-the-end iterator.
> 
> Objects involved in the operation:
>     iterator "this" @ 0x0x7ffd73f035b0 {
>       type = gnu_debug::_Safe_iterator<std::_Rb_tree_iterator<std::pair<uint256 const, CMasternodeBroadcast> >, std::debug::map<uint256, CMasternodeBroadcast, std::less<uint256>, std::allocator<std::pair<uint256 const, CMasternodeBroadcast> > > > (mutable iterator);
>       state = past-the-end;
>       references sequence with type 'std::__debug::map<uint256, CMasternodeBroadcast, std::less<uint256>, std::allocator<std::pair<uint256 const, CMasternodeBroadcast> > >' @ 0x0x55a98bf6ae20
>     }
> ```
> Before changing dependencies this error did not happen to me. Can anyone give me a hint to resolve this issue? It happend running on Ubuntu 16.04 with testing libstdc++ libs from apt-repository ppa:ubuntu-toolchain-r/test to get on the actual version for 16.04:
> 
> ```
> add-apt-repository ppa:ubuntu-toolchain-r/test
> apt-get update && apt-get install libstdc++6
> ```
> Thanks!

@flyinghuman yes, Bionic introduces a new libstdc++. you would need to configure/compile with the `--enable-glibc-back-compat` option for resulting binary compatibility on 16.04 and below.

Gitian and TravisCI builds do this already

**Comment by flyinghuman** [comment]
even with enable-glibc-back-compat the error occurs. i now made a
```
LDFLAGS='-static-libstdc++ -static-libgcc' CFLAGS=-fPIC
```
to include the lib - this seems to work now.
However, with the actual depends i get strange errors if i let compile the full QT-depends (remove NO_QT=1). It says that it can't find the Freetype headers/functions in QT. Any hint for this issue?
(See the attached image).. some sort of linking issue.
I will try on new virtual machine to reproduce the error.

Greets, Fly.
image

**Comment by Warrows** [comment]
ACK 1c4db75. Works on debian 8 with libstdc++ 4.9.2, which was failing previously.

**Comment by Fuzzbawls** [comment]
the force push with 1c4db75 as the HEAD fixes some linker issues and a missing header include that is required on some systems.

@flyinghuman: regarding the depends build issue, what host OS are you running the depends build on?

**Comment by flyinghuman** [comment]
i tried on travis with ubuntu 18.04 docker. without compiling full QT and using system QT it works.

**Comment by flyinghuman** [comment]
i recommend to add the -k tag to curl because as of today, gmp cannot be downloaded because of invalid SSL-Certificate from thier Server. with the -k tag curl ignores the Certificate: https://github.com/flyinghuman/PIVX/commit/608bcd386a4f789e00e9924d2e1da4126b14a719

**Comment by Fuzzbawls** [comment]
libgmp.org's SSL cert has been updated and is now valid until the year 2022 so the downloading issue has resolved itself.

That said, I would prefer using alternate mirrors for package downloads before using the `-k` option for curl. Up until the addition of GMP as a dependency, the upstream fallback URL for package downloads (bitcoincore.org) has been sufficient. We could offer full redundancy by setting up our own fallback mirror for all dependency packages though.

Regarding the macOS graphic glitches: reproduction of the issue has not happened; none of my macs have ever shown the aforementioned glitch over numerous builds. As such, I don't think this PR should be held back because of it. The rationale here is that the graphical glitch is not actually related to the build system changes proposed here. A new issue report and subsequent PR would be the more appropriate place to discuss/test/fix such an issue, especially given it's apparent intermittent reproducibility.

**Pull request merged**